### PR TITLE
New style PReLU

### PR DESCRIPTION
**Pull request by hvy**
New style Parametric ReLU (PReLU) to support double backprop.

### CODE REVIEW

**Comment by niboshi** [comment]
Please rebase.

**Comment by hvy** [comment]
@niboshi Rebased and updated tolerances.

File: chainer/functions/activation/prelu.py [hunk 1]
```diff
@@ -32,45 +33,90 @@ def forward_cpu(self, inputs):
...
 
+    def backward(self, indexes, grad_outputs):
+        x, W, gy = self.get_retained_inputs()
+        ggx, ggW = grad_outputs
+
+        ggW = chainer.functions.broadcast_to(
+            chainer.functions.reshape(ggW, self.extended_shape), x.shape)
+        ggW *= self.cond < 0
+
+        gxgy, gxW = PReLUFunctionGrad(self.cond, None, self.extended_shape) \
```
Comment (niboshi) [inline chainer/functions/activation/prelu.py]: Ditto. (`\`)

File: chainer/functions/activation/prelu.py [hunk 2]
```diff
@@ -32,45 +33,90 @@ def forward_cpu(self, inputs):
...
+            .apply((gy, W, ggx))
+
+        return gy * ggW, chainer.functions.sum(gxW, axis=self.reduce_axes), \
```
Comment (niboshi) [inline chainer/functions/activation/prelu.py]: Ditto. (`\`)

File: chainer/functions/activation/prelu.py [hunk 3]
```diff
@@ -32,45 +33,90 @@ def forward_cpu(self, inputs):
...
+            x_type.shape[1:1 + type_check.eval(W_type.ndim)] == W_type.shape,
+            gy_type.shape == x_type.shape
+        )
+
+    def forward_cpu(self, inputs):
+        x, W, gy = inputs
+        mask = self.cond >= 0
+        masked = numpy.where(mask, 0, x * gy)
+
+        gW = masked if self.reduce_axes is None else \
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/activation/prelu.py]: Changed to regular if statements and added comments.

File: chainer/functions/activation/prelu.py (hunk 2 above)
Comment (hvy) [reply chainer/functions/activation/prelu.py]: Shortened line lengths by returning variable length gradient list instead wrt `indexes`.

File: chainer/functions/activation/prelu.py (hunk 1 above)
Comment (hvy) [reply chainer/functions/activation/prelu.py]: Wrapped in `()`

File: chainer/functions/activation/prelu.py [hunk 4]
```diff
@@ -32,45 +33,90 @@ def forward_cpu(self, inputs):
...
-            'T x, T gy', 'T masked',
-            'masked = x >= 0 ? (T)0 : (T)(x * gy)',
-            'prelu_masked')(x, gy)
-        axes = (0,) + tuple(six.moves.range(1 + W.ndim, gy.ndim))
-        gW = masked.sum(axis=axes)
+            'T x, T cond, T gy', 'T masked',
+            'masked = cond >= 0 ? (T)0 : (T)(x * gy)',
+            'prelu_masked')(x, self.cond, gy)
+
+        gW = masked.copy() if self.reduce_axes is None else \
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/activation/prelu.py]: Changed to regular if statements.

File: chainer/functions/activation/prelu.py [hunk 5]
```diff
@@ -102,8 +148,12 @@ def prelu(x, W):
...
-    return PReLUFunction()(x, W)
+    return PReLUFunction().apply((x, W))[0]
 
 
 def _get_extended_shape(W, x):
     return (1,) + W.shape + (1,) * (x.ndim - W.ndim - 1)
+
+
+def _get_reduce_axes(W, x):
+    return (0,) + tuple(six.moves.range(1 + W.ndim, x.ndim))
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/activation/prelu.py]: Removed six dependency.

**Comment by hvy** [comment]
Thank you again for the review. I updated the code accordingly.

**Comment by niboshi** [comment]
Thank you!

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**